File: src/lib/cheese-home.ts
import { execFileSync } from "node:child_process";
import { mkdirSync, readFileSync, realpathSync, writeFileSync } from "node:fs";
import os from "node:os";
import path from "node:path";

export interface CheeseHomePaths {
  root: string;
  projectDir: string;
  milknadoDb: string;
  worktreeDir: string;
  manifestsDir: string;
  runsDir: string;
  sharedDir: string;
}

export interface CheeseHomeOptions {
  home?: string;
  canonicalRepo?: string;
}

export interface RetentionConfig {
  defaultDays: number;
  milknadoDays?: number;
  manifestsDays?: number;
  runsDays?: number;
  worktreeDays?: number;
}

const DEFAULT_RETENTION_DAYS = 30;
const RETENTION_NUMERIC_KEYS: ReadonlySet<keyof RetentionConfig> = new Set([
  "defaultDays",
  "milknadoDays",
  "manifestsDays",
  "runsDays",
  "worktreeDays",
]);

export function pathSlug(absPath: string): string {
  return absPath.replaceAll("/", "-");
}

export function discoverCanonicalRepo(cwd: string): string {
  const out = runGitWorktreeList(cwd);
  return realpathSync(parseWorktreeMain(out, cwd));
}

export function parseWorktreeMain(out: string, cwd: string): string {
  const firstLine = out.split("\n", 1)[0] ?? "";
  if (!firstLine.startsWith("worktree ")) {
    throw new Error(
      `discoverCanonicalRepo: ${cwd} is not inside a git worktree (no 'worktree' line)`,
    );
  }
  return firstLine.slice("worktree ".length);
}

function runGitWorktreeList(cwd: string): string {
  try {
    return execFileSync("git", ["worktree", "list", "--porcelain"], {
      cwd,
      encoding: "utf8",
      stdio: ["ignore", "pipe", "pipe"],
    });
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(
      `discoverCanonicalRepo: ${cwd} is not inside a git repo: ${message}`,
    );
  }
}

export function resolveCheeseHome(
  cwd: string,
  options: CheeseHomeOptions = {},
): CheeseHomePaths {
  const root = options.home ?? path.join(os.homedir(), ".cheese");
  const canonicalRepo = options.canonicalRepo ?? discoverCanonicalRepo(cwd);
  const repoSlug = pathSlug(canonicalRepo);
  const worktreePath = realpathSync(cwd);
  const wtSlug = pathSlug(worktreePath);
  const projectDir = path.join(root, "projects", repoSlug);
  const worktreeDir = path.join(projectDir, "worktrees", wtSlug);
  return {
    root,
    projectDir,
    milknadoDb: path.join(projectDir, "milknado", "milknado.db"),
    worktreeDir,
    manifestsDir: path.join(worktreeDir, "manifests"),
    runsDir: path.join(worktreeDir, "runs"),
    sharedDir: path.join(projectDir, "shared"),
  };
}

export function ensureCheeseHome(
  cwd: string,
  options: CheeseHomeOptions = {},
): CheeseHomePaths {
  const paths = resolveCheeseHome(cwd, options);
  mkdirSync(path.dirname(paths.milknadoDb), { recursive: true });
  mkdirSync(paths.manifestsDir, { recursive: true });
  mkdirSync(paths.runsDir, { recursive: true });
  mkdirSync(paths.sharedDir, { recursive: true });
  writeSidecar(paths.worktreeDir, realpathSync(cwd));
  return paths;
}

function writeSidecar(worktreeDir: string, originalPath: string): void {
  writeFileSync(path.join(worktreeDir, ".path"), `${originalPath}\n`, "utf8");
}

export function readRetentionConfig(projectDir: string): RetentionConfig {
  const tomlPath = path.join(projectDir, "shared", "retention.toml");
  let body: string;
  try {
    body = readFileSync(tomlPath, "utf8");
  } catch {
    return { defaultDays: DEFAULT_RETENTION_DAYS };
  }
  return parseRetentionToml(body);
}

function parseRetentionToml(body: string): RetentionConfig {
  const config: Record<string, number> = {
    defaultDays: DEFAULT_RETENTION_DAYS,
  };
  for (const rawLine of body.split(/\r?\n/)) {
    const line = stripComment(rawLine).trim();
    if (line.length === 0 || line.startsWith("[")) {
      continue;
    }
    const eq = line.indexOf("=");
    if (eq < 0) continue;
    const key = line.slice(0, eq).trim();
    const value = line.slice(eq + 1).trim();
    if (!RETENTION_NUMERIC_KEYS.has(key as keyof RetentionConfig)) continue;
    const parsed = Number.parseInt(value, 10);
    if (Number.isFinite(parsed)) {
      config[key] = parsed;
    }
  }
  return config as unknown as RetentionConfig;
}

function stripComment(line: string): string {
  const hash = line.indexOf("#");
  return hash < 0 ? line : line.slice(0, hash);
}

File: python/cheese_flow/lib/cheese_home.py
"""Port of `src/lib/cheese-home.ts` — ``~/.cheese`` home directory resolution.

Mirrors the TS surface: ``CheeseHomePaths``, ``CheeseHomeOptions``,
``RetentionConfig``, ``path_slug``, ``discover_canonical_repo``,
``parse_worktree_main``, ``resolve_cheese_home``, ``ensure_cheese_home``,
and ``read_retention_config``.

Field names on the dataclasses keep TS camelCase (``projectDir``,
``milknadoDb``, etc.) so callers ported from TS read the same.
"""

from __future__ import annotations

import os
import subprocess
from dataclasses import dataclass
from pathlib import Path

DEFAULT_RETENTION_DAYS = 30
_RETENTION_NUMERIC_KEYS: frozenset[str] = frozenset(
    {
        "defaultDays",
        "milknadoDays",
        "manifestsDays",
        "runsDays",
        "worktreeDays",
    }
)


@dataclass(frozen=True)
class CheeseHomePaths:
    root: str
    projectDir: str
    milknadoDb: str
    worktreeDir: str
    manifestsDir: str
    runsDir: str
    sharedDir: str


@dataclass(frozen=True)
class CheeseHomeOptions:
    home: str | None = None
    canonicalRepo: str | None = None


@dataclass
class RetentionConfig:
    defaultDays: int = DEFAULT_RETENTION_DAYS
    milknadoDays: int | None = None
    manifestsDays: int | None = None
    runsDays: int | None = None
    worktreeDays: int | None = None


def path_slug(abs_path: str) -> str:
    return abs_path.replace("/", "-")


def parse_worktree_main(out: str, cwd: str) -> str:
    first_line = out.split("\n", 1)[0] if out else ""
    if not first_line.startswith("worktree "):
        raise RuntimeError(
            f"discoverCanonicalRepo: {cwd} is not inside a git worktree (no 'worktree' line)"
        )
    return first_line[len("worktree ") :]


def _run_git_worktree_list(cwd: str) -> str:
    try:
        result = subprocess.run(
            ["git", "worktree", "list", "--porcelain"],
            cwd=cwd,
            capture_output=True,
            text=True,
            check=True,
        )
    except (subprocess.CalledProcessError, FileNotFoundError) as error:
        message = str(error)
        raise RuntimeError(
            f"discoverCanonicalRepo: {cwd} is not inside a git repo: {message}"
        ) from error
    return result.stdout


def discover_canonical_repo(cwd: str) -> str:
    out = _run_git_worktree_list(cwd)
    return str(Path(parse_worktree_main(out, cwd)).resolve())


def resolve_cheese_home(cwd: str, options: CheeseHomeOptions | None = None) -> CheeseHomePaths:
    opts = options or CheeseHomeOptions()
    root = opts.home if opts.home is not None else str(Path.home() / ".cheese")
    canonical_repo = (
        opts.canonicalRepo if opts.canonicalRepo is not None else discover_canonical_repo(cwd)
    )
    repo_slug = path_slug(canonical_repo)
    worktree_path = str(Path(cwd).resolve())
    wt_slug = path_slug(worktree_path)
    project_dir = os.path.join(root, "projects", repo_slug)
    worktree_dir = os.path.join(project_dir, "worktrees", wt_slug)
    return CheeseHomePaths(
        root=root,
        projectDir=project_dir,
        milknadoDb=os.path.join(project_dir, "milknado", "milknado.db"),
        worktreeDir=worktree_dir,
        manifestsDir=os.path.join(worktree_dir, "manifests"),
        runsDir=os.path.join(worktree_dir, "runs"),
        sharedDir=os.path.join(project_dir, "shared"),
    )


def ensure_cheese_home(cwd: str, options: CheeseHomeOptions | None = None) -> CheeseHomePaths:
    paths = resolve_cheese_home(cwd, options)
    os.makedirs(os.path.dirname(paths.milknadoDb), exist_ok=True)
    os.makedirs(paths.manifestsDir, exist_ok=True)
    os.makedirs(paths.runsDir, exist_ok=True)
    os.makedirs(paths.sharedDir, exist_ok=True)
    _write_sidecar(paths.worktreeDir, str(Path(cwd).resolve()))
    return paths


def _write_sidecar(worktree_dir: str, original_path: str) -> None:
    Path(worktree_dir, ".path").write_text(f"{original_path}\n", encoding="utf-8")


def read_retention_config(project_dir: str) -> RetentionConfig:
    toml_path = Path(project_dir) / "shared" / "retention.toml"
    try:
        body = toml_path.read_text(encoding="utf-8")
    except OSError:
        return RetentionConfig()
    return _parse_retention_toml(body)


def _strip_comment(line: str) -> str:
    hash_idx = line.find("#")
    return line if hash_idx < 0 else line[:hash_idx]


def _parse_retention_toml(body: str) -> RetentionConfig:
    config: dict[str, int] = {"defaultDays": DEFAULT_RETENTION_DAYS}
    for raw_line in body.splitlines():
        line = _strip_comment(raw_line).strip()
        if line == "" or line.startswith("["):
            continue
        eq = line.find("=")
        if eq < 0:
            continue
        key = line[:eq].strip()
        value = line[eq + 1 :].strip()
        if key not in _RETENTION_NUMERIC_KEYS:
            continue
        try:
            parsed = int(value, 10)
        except ValueError:
            continue
        config[key] = parsed
    return RetentionConfig(
        defaultDays=config.get("defaultDays", DEFAULT_RETENTION_DAYS),
        milknadoDays=config.get("milknadoDays"),
        manifestsDays=config.get("manifestsDays"),
        runsDays=config.get("runsDays"),
        worktreeDays=config.get("worktreeDays"),
    )
